ddl/placement/bundle_test.go

```diff
@@ -553,14 +547,442 @@ func (s *testBundleSuite) TestString(c *C) {
+		name: "sugar syntax: normal case 2",
+		input: &model.PlacementSettings{
+			PrimaryRegion: "us",
+			Regions:       "us,sh",
+			Followers:     5,
+		},
+		output: []*Rule{
+			{
+				Role: Leader,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 1,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 3,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"sh"},
+					},
+				},
+				Count: 2,
+			},
+		},
+	})
+	tests = append(tests, tests[len(tests)-1])
+	tests[len(tests)-1].name = "sugar syntax: explicit schedule"
+	tests[len(tests)-1].input.Schedule = "even"
+
+	tests = append(tests, TestCase{
+		name: "sugar syntax: majority schedule",
+		input: &model.PlacementSettings{
+			PrimaryRegion: "us",
+			Regions:       "bj,sh",
+			Followers:     5,
+			Schedule:      "majority_in_primary",
+		},
+		output: []*Rule{
+			{
+				Role: Leader,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 1,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 3,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"bj"},
+					},
+				},
+				Count: 1,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"sh"},
+					},
+				},
+				Count: 1,
+			},
+		},
+	})
+
+	tests = append(tests, TestCase{
+		name: "direct syntax: normal case 1",
+		input: &model.PlacementSettings{
+			Constraints: "[+region=us]",
+			Followers:   2,
+		},
+		output: []*Rule{
+			{
+				Role: Leader,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 1,
+			},
+			{
+				Role: Follower,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 2,
+			},
+		},
+	})
+
+	tests = append(tests, TestCase{
+		name: "direct syntax: normal case 2",
+		input: &model.PlacementSettings{
+			Constraints: "[+region=us]",
+			Voters:      2,
+		},
+		output: []*Rule{
+			{
+				Role: Leader,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 1,
+			},
+			{
+				Role: Voter,
+				Constraints: Constraints{
+					{
+						Key:    "region",
+						Op:     In,
+						Values: []string{"us"},
+					},
+				},
+				Count: 2,
```

Comment:
I got a schedule problem here.

For leader, the rule1 is suitable, if the rule2 is to schedule voters to `cn`, does leader should conform this?

Reply:
PD will handle this. Whatever the result will be, we don't handle it.

We have discussed similar questions in the slack, https://tidbcommunity.slack.com/archives/C02BZE39NBU/p1630063091019900. Since all conditions will be confirmed, voters will become an alias of followers.

`LEADER_CONSTRAINTS="{+region=sh:1}" VOTER_CONSTRAINTS="{+region=bj:2} FOLLOWER_CONSTRAINTS="{+region=hz:3}"`, @lcwangchao had proposed an example.

Maybe this syntax is too confusing? `voter` is clearly useful for the implementation of PD, but not friendly enough for users(I need to check the source code to confirm the behavior). Maybe we should provide `VOTERS_CONSTRAINTS=[list]`, just like `CONSTRAINTS` to be easier to reason about?

Reply:
I'm thinking about wether `LEADER_CONSTRAINTS` + `VOTER_CONSTRAINTS` is necessary, maybe the combination `LEADER_CONSTRAINTS` + `FOLLOWER_CONSTRAINTS ` or `VOTER_CONSTRAINTS` + `FOLLOWER_CONSTRAINTS` is enough

Reply:
`LEADER_CONSTRAINTS ` is clearly needed in some cases. I think we could discuss this in the next meeting or in the slack. IMO, I would like to allow only `VOTERS_CONSTRAINTS=[list]`, then remove `VOTERS=xxx`.